sycl/doc/DeviceGlobal.md

```diff
@@ -0,0 +1,617 @@
+quite different from device global variables that are not declared with this
+property.  The sections below describe both implementations.
+
+
+## Design
+
+### Changes to DPC++ headers
+
+#### Class specializations based on `device_image_scope`
+
+The headers, of course, include the declaration of the new `device_global`
+class, which is described in the [extension specification][1].  The declaration
+of this class uses partial specialization to define the class differently
+depending on whether it has the `device_image_scope` property.  When the
+property is not present, the class has a member variable which is a pointer to
+the underlying type.  Member functions which return a reference to the value
+(e.g. `get`) return the value of this pointer:
+
+```
+template<typename T, typename PropertyListT>
+class device_global {
+  T *usmptr;
+ public:
+  T& get() noexcept { return *usmptr; }
+  /* other member functions */
+};
+```
+
+However, when the property is present, it has a member variable which is the
+type itself, and member functions return a reference to this value.
+
+```
+template<typename T, typename PropertyListT>
+class device_global {
+  T val;
+ public:
+  device_global() : val() {}
+  T& get() noexcept { return val; }
+  /* other member functions */
+};
+```
+
+Note that the default constructor "value initializes" the `val` member
+variable.  Since the type `T` is limited to types that are trivially
+constructible, this means that `val` will be zero initialized.
+
+In both cases the member variable (either `usmptr` or `val`) must be the first
+member variable in the class.  As we will see later, the runtime assumes that
+the address of the `device_global` variable itself is the same as the address
+of this member variable.
+
+#### Attributes attached to the class
+
+The `device_global` class declaration contains three C++ attributes which
+convey information to the front-end.  These attributes are only needed for the
+device compiler, and the `#ifdef __SYCL_DEVICE_ONLY__` allows the customer to
+use another host compiler, even if it does not recognize these attributes.
+Also note that these attributes are all in the `__sycl_detail__` namespace, so
+they are considered implementation details of DPC++.  We do not intend to
+support them as general attributes that customer code can use.
+
+```
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+// Partial specialization to make PropertyListT visible as a parameter pack
+// of properties.
+template <typename T, typename ...Props>
+class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_global_variable_attributes(
+    "sycl-device-global-size",
+    Props::meta_name...,
+    sizeof(T),
+    Props::meta_value...
+    )]]
+  [[__sycl_detail__::global_variable_allowed]]
+  [[__sycl_detail__::restrictions(device_global)]]
+#endif
+  device_global<T, property_list<Props...>> {/*...*/};
+```
```

Comment:
Just want to check my understanding here.  Does the use of the attributes depend on the use of a parameter pack?  If so, does that mean that a declaration like this `device_global<int, property_list_t<some_property::value_t>>  dgi;` would NOT get the benefit of the attributes, because the compiler will  select the more exact property list specialization rather than the parameter pack one?   Or should all specializations receive the attributes?

And, we have now four full specializations of device_global, correct?  We are specializing it for with and without `device_image_scope` (ie `usmptr` vs `val` ) and both of those for `PropertyListT` and parameter pack?

Reply:
> Just want to check my understanding here. Does the use of the attributes depend on the use of a parameter pack? If so, does that mean that a declaration like this `device_global<int, property_list_t<some_property::value_t>> dgi;` would NOT get the benefit of the attributes, because the compiler will select the more exact property list specialization rather than the parameter pack one? Or should all specializations receive the attributes?

The intent is that all device global declarations should use the specialization that defines the attribute.  Now that you mention it, though, maybe this partial specialization should be this instead:

```
device_global<T, property_list_t<Props...>> {/*...*/};
```

It so happens that we are rethinking some of the details of `property_list`, which will also affect how these partial specializations work.  I'll work with @rolandschulz to update the design doc to use the new property list syntax and also make sure this partial specialization is correct.  That update will come once we finalize our updates to property list, hopefully soon.

> And, we have now four full specializations of device_global, correct? We are specializing it for with and without `device_image_scope` (ie `usmptr` vs `val` ) and both of those for `PropertyListT` and parameter pack?

Yes, that's correct.  We will need partial specializations for both the case when `device_image_scope` is present and when `device_image_scope` is not present.

Reply:
I wonder if we should just have one representation for the data. A `val` followed by a `usmptr` and use that everywhere.  Then we can use `enable_if` on individual methods to differentiate `device_image_scope` from not when deciding which member variable to update and modify?

Reply:
That will allocate a lot of unused memory in the case when `device_image_scope` is not specified and when the size of `T` is large.  In that case, we will allocate a large `val` which will never be used, and this unused space will be wasted on the host and every device that uses the device global variable.

Reply:
Since you have a great design document, be sure to captures some of the new motivations/explanations/clarifications from the questions/answers.

Reply:
Good idea.  Added in ff903a312d98d576cb3c443f26d71c8e942b7ec6.